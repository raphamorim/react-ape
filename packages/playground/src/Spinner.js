function Spinner(props, apeContext) {
  const {ctx, next} = apeContext;
  const {style = {}, degrees} = props;
  console.log('spinner', degrees, apeContext)

  const offset = 8;
  ctx.save();
  ctx.translate(offset, offset);
  ctx.rotate(degrees);

  // Draw half open circle
  ctx.beginPath();
  ctx.lineWidth = 2;
  ctx.arc(8 - offset, 8 - offset, 6, 0, 1.75 * Math.PI);
  ctx.strokeFill = props.style.color;
  ctx.stroke();

  // Draw arrowhead
  ctx.lineWidth = 2;
  ctx.moveTo(13 - offset, 1 - offset);
  ctx.lineTo(9 - offset, 5 - offset);
  ctx.lineTo(13 - offset, 5 - offset);
  ctx.lineTo(13 - offset, 1 - offset);
  ctx.stroke();
  ctx.restore();
}

export default Spinner;
