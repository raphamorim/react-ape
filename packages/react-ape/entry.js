/**
 * Copyright (c) 2018-present, Raphael Amorim.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import ReactApeRenderer from './renderer/renderer';
import StyleSheetModule from './modules/StyleSheet';
import DimensionsModule from './modules/Dimensions';
import PlatformModule from './modules/Platform';

import ListViewComponent from './renderer/components/ListView';
import RegisterComponentFn from './modules/Register';

import withFocusFn from './modules/Navigation/withFocus';
import withNavigationFn from './modules/Navigation/withNavigation';

export const ListView = ListViewComponent;
export const registerComponent = RegisterComponentFn;

export const render = ReactApeRenderer.render;
// export const unmountComponentAtNode = ReactApeRender.unmountComponentAtNode;

export const Image = 'Image';
export const View = 'View';
export const Text = 'Text';

export const StyleSheet = StyleSheetModule;
export const Dimensions = DimensionsModule;
export const Platform = PlatformModule;
export const withFocus = withFocusFn;
export const withNavigation = withNavigationFn;

export default ReactApeRenderer;
