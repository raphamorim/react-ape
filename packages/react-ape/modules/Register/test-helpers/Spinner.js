class Spinner {
  reset(prevProps, parentStyle, canvas) {
    const {ctx} = canvas;
    // parentStyle.backgroundColor // white
    if (ctx) {
      ctx.clearRect(0, 0, 18, 18);
    }
  }

  render(props, canvas) {
    const {ctx} = canvas;
    const {style = {}, degrees} = props;
    const {color = 'black'} = style;

    const offset = 8;
    ctx.save();
    ctx.translate(offset, offset);
    ctx.rotate(degrees);

    // Draw half open circle
    ctx.beginPath();
    ctx.lineWidth = 2;
    ctx.arc(8 - offset, 8 - offset, 6, 0, 1.75 * Math.PI);
    ctx.strokeStyle = color;
    ctx.stroke();

    // Draw arrowhead
    ctx.lineWidth = 2;
    ctx.moveTo(13 - offset, 1 - offset);
    ctx.lineTo(9 - offset, 5 - offset);
    ctx.lineTo(13 - offset, 5 - offset);
    ctx.lineTo(13 - offset, 1 - offset);
    ctx.stroke();
    ctx.restore();
  }
}

const spinner = new Spinner();
export default spinner;
