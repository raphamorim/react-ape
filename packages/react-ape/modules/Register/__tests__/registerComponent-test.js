import React, {useState, useEffect} from 'react';

import registerComponent, {CustomComponents} from '../index';
import Spinner from '../test-helpers/Spinner';

import {render, View, Text, StyleSheet} from '../../../entry';

function testCanvasSnapshot(expect, canvas) {
  expect(canvas.toDataURL()).toMatchSnapshot();
}

describe('registerComponent', () => {
  describe('registerComponent("ComponentName", Callable)', () => {
    test('should register component and return "ComponentName"', () => {
      const canvas = document.createElement('canvas');
      expect(Object.keys(CustomComponents).length).toEqual(0);
      expect(typeof registerComponent).toEqual('function');
      registerComponent('Spinner', Spinner);
      expect(Object.keys(CustomComponents).length).toEqual(1);

      const custom = {Spinner: 'Spinner'};
      function App() {
        return (
          <View style={{backgroundColor: 'white'}}>
            <custom.Spinner degrees={0.1} style={{color: 'blue'}} />
          </View>
        );
      }

      render(<App />, canvas, () => testCanvasSnapshot(expect, canvas));
    });

    test.skip('should update when props change', done => {
      const canvas = document.createElement('canvas');
      expect(typeof registerComponent).toEqual('function');
      registerComponent('Spinner', Spinner);

      const custom = {Spinner: 'Spinner'};
      class AppWithUpdate extends React.Component {
        constructor(props) {
          super(props);
          this.state = {degrees: 0.0};
        }

        componentDidMount() {
          setTimeout(() => {
            const {degrees} = this.state;
            this.setState({degrees: 0.1}, () => {
              testCanvasSnapshot(expect, canvas);
              done();
            });
          });
        }

        render() {
          const {degrees} = this.state;
          return (
            <View>
              <custom.Spinner degrees={degrees} style={{color: 'green'}} />
            </View>
          );
        }
      }

      render(<AppWithUpdate />, canvas, () =>
        testCanvasSnapshot(expect, canvas)
      );
    });
  });
});
