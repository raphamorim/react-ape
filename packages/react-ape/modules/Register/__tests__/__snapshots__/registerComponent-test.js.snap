// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerComponent registerComponent("ComponentName", Callable) should register component and return "ComponentName" 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAD3UlEQVR4nO3cu2pUaxzG4XcFC2MteGjsLazUykoQCwsFL8BWBDVEO69ARDzgqbSyV6xEvQIrEa9AIZjUKiKZXSzcMSbitkgy787zwDAzi/XBHwZ+fDOzZobJZDLJX/j2LVleTnbu/JtV0GkYhmGrZ2DFzK8H3r5Njh1LZmeTYRhvu3aNx169So4fT86eTV6/3opxge1s+HmHNTeX3L2b/Jc914ULyf37GzkabD07rOny7w5rbi65c2eM1d694+PFxfF2+vTahfv3b+aYAMmOJHn3btxZJcmpU8nz5+Pj5eXk2rXk2bO1CwUL2Gw7kuT8+ZWd1Y9Yff+enDuXPHmy/sJ9+zZrRIDRMJlMJrOzydev49vAS5dWn/D5c/LxY7KwsHL/4UNy8WJy4MDWDA2bxWdY02WYTCaTHy/J4mKye/fWDgTTRLCmy6rLGpaX/7xgaSk5cyaZn9+okQDWN5OM11wlv/+86mePHydPnyaPHm3gVADrmEmSw4fHJ9ev/3nBjRvj/dGjGzUSwPpmkuTevfGK9oWF5OTJ35984kTy6dN47oMHmzUiwGgmSQ4dSi5fHg+8eDFe3nDr1hinpaXk5s1kz57k5cvxnKtXk4MHt2pkYLta9dOc+fnk9u3f/zRnGJIrV1beFsL/nW8Jp8vw6781vH8/Xkj65k3y5ct4bHY2OXIkefjQzortRbCmy5pgASsEa7qs+XsZgGklWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1BAsoIZgATUEC6ghWEANwQJqCBZQQ7CAGoIF1PgHCyaKAtUF30MAAAAASUVORK5CYII="`;

exports[`registerComponent registerComponent("ComponentName", Callable) should update when props change 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAACSUlEQVR4nO3bv4oVZxjH8e+sBhIQIUTWW5CgIRjL3EHuQEidBEGjYBCbPVsIWvinSJpcQEiROoRAriBJm1tQ06RIoeC6k+LM2XXXjYvgnlP4+VTzzjwv/KqH4Zl36nXMWm+zr15rD8AbMrx0Z9aH1ffVJ429O1U9qf6sTlVnprqX9wIcoeN7Vhvdb+hK49SMdlvSe9WnywwGsN/aztVmD6qvp2b1qLWudLIPOtaZ6p9VBQRYmL9h3erjtro8vVH93KzPqvnMqn6q3l9RPoAd84b1rO+qoaGHbexpVr9VZ1eWDuAFixnW+aqGbu88mfV3dW75kQAOtphhzb8GbvfD6qIAvNrxw0v2udXpnvWoobGNF4b2AEds0XCeTquLh+7YnmrGaQ/Akiwa1h9VbXfz0B1b3Ziufj+aSAAHWzSsLxoaq9PN+uV/q2f92tB6Q2PvdGkpCQEmu2fZN7vX2NVp9bhj3elEP7bV8/7t84a+aWi9sVrrfhtdW0li4K2193/Aze61e9r9IGN1t1nXjzoYwH4H/fz8UfVtdaH5cYehetJ8zvVls/5aZkAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgBX5DwSNQ801188SAAAAAElFTkSuQmCC"`;

exports[`registerComponent registerComponent("ComponentName", Callable) should update when props change 2`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAACfElEQVR4nO3bMYsdVRjH4d/sLrLB1oSF+AlEbGLnVxDttdBGUJEkCMZyZ4sgIlErSWFj4QewsBDFD6BiqZ1YaQQbi7BFdMdi7q675i4SZCfFPg/M5Zw754V/9TJzmFMPauyRxrYfuA7gfxru+2fsqeqj6kq13VQN7VffN7Tb1F71R3Wrsa+XDAucb1snZnt92NTVjjeyeXSheqapr46t/rk0LGA5G0ejuVlda25Rd9roWnVxdX12X+XQLwtlBKgOn7Bu9mT3ulrV0Oft9mxVYxvVzeq5NbUaFrCouWHd63Y1NHWn8ahZbVWfVC+srZz6dZmIALPDPawrVW32ztGdsT+rFxt7pc0uN7XTQZernerxtvpx6bDA+TZvqY9Nq/nFxn5/eHEATjc/Yc2fLtSjxzbhTzP2WEMfVz+125tnmg7gmI3V735Vd0/ZrzpZ8XJTzzf16pkmA/iXuWFNfbeav/2fFX/11mr0zdlEAljv8BXwjeYXw53Gvjh19diXDV1qaKpeXyAfwJF/vmjf7YOGrq9mv7XZu233aXc76KCXGrrR0KXV/fcau7F4WuBcO3mWcK/3q+tNa84Yzqunpm41Hr0WAixm3eHnJ6rbDT3dQRdWq/arb6vXGvth0YQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/H3/8nVaIB2iTpAAAAAElFTkSuQmCC"`;
